### PR TITLE
chore(env): updated the env getter function & default jobCleanUpPolicy as retain

### PR DESCRIPTION
**Pull request by ispeakc0de**
Signed-off-by: shubhamchaudhary <shubham@chaosnative.com>

**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

-  Marked default jobCleanupPolicy policy as retain
- updated the env getter function

**Checklist:**
- [ ] Fixes #<issue number>
- [ ] Labelled this PR & related issue with `documentation` tag
- [ ] PR messages has document related information
- [ ] Labelled this PR & related issue with `breaking-changes` tag
- [ ] PR messages has breaking changes related information
- [ ] Labelled this PR & related issue with `requires-upgrade` tag
- [ ] PR messages has upgrade related information
- [ ] Commit has unit tests
- [ ] Commit has integration tests

### CODE REVIEW

**Comment by ksatchit** [comment]
Just added @kazukousen for review - as author of a recent refactor of the same!

**Review by kazukousen — DISMISSED** [review]
> updated the env getter function

looks good to me, thanks

File: pkg/utils/environment.go [hunk 1]
```diff
@@ -38,6 +38,20 @@ func (engineDetails *EngineDetails) SetEngineUID(clients ClientSets) error {
 
 //SetENV sets ENV values in experimentDetails struct.
 func (expDetails *ExperimentDetails) SetENV(engineDetails EngineDetails, clients ClientSets) error {
+	// Setting envs from engine fields other than env
+	expDetails.setEnv("CHAOSENGINE", engineDetails.Name).
```
Comment (rahulchheda) [inline pkg/utils/environment.go:42]: nit: Do we really need builder for setting ENV's. I would suggest a slice of ENV variables, as in `v1.EnvVar`, and then we don't have to do same function call again and again.

File: pkg/utils/watchJob.go [hunk 2]
```diff
@@ -99,16 +101,24 @@ func (engineDetails EngineDetails) DeleteJobAccordingToJobCleanUpPolicy(experime
 		return "", err
 	}
 
-	if expEngine.Spec.JobCleanUpPolicy == v1alpha1.CleanUpPolicyDelete || string(expEngine.Spec.JobCleanUpPolicy) == "" {
-		log.Infof("deleting the job as jobCleanPolicy is set to %v", expEngine.Spec.JobCleanUpPolicy)
+	jobCleanUpPolicy := v1alpha1.CleanUpPolicyRetain
```
Comment (rahulchheda) [inline pkg/utils/watchJob.go:104]: no need for this variable

File: pkg/utils/watchJob.go [hunk 3]
```diff
@@ -99,16 +101,24 @@ func (engineDetails EngineDetails) DeleteJobAccordingToJobCleanUpPolicy(experime
...
+	if expEngine.Spec.JobCleanUpPolicy != "" {
```
Comment (rahulchheda) [inline pkg/utils/watchJob.go:105]: use this instead for  the switch case.

File: pkg/utils/watchJob.go [hunk 4]
```diff
@@ -99,16 +101,24 @@ func (engineDetails EngineDetails) DeleteJobAccordingToJobCleanUpPolicy(experime
...
+		if deleteJob := clients.KubeClient.BatchV1().Jobs(experiment.Namespace).Delete(experiment.JobName, &metav1.DeleteOptions{
 			PropagationPolicy: &deletePolicy,
-		})
-		if deleteJob != nil {
+		}); deleteJob != nil {
 			return "", errors.Errorf("unable to delete ChaosExperiment Job name: %v, in namespace: %v, error: %v", experiment.JobName, experiment.Namespace, err)
 		}
+	case v1alpha1.CleanUpPolicyRetain:
+		log.Infof("[skip]: skipping the job deletion as jobCleanUpPolicy is set to {%v}", expEngine.Spec.JobCleanUpPolicy)
+	default:
```
Comment (rahulchheda) [inline pkg/utils/watchJob.go:120]: handling for `string(expEngine.Spec.JobCleanUpPolicy) == ""` is already in default, no need to create new variable for this.

File: pkg/utils/watchJob.go [hunk 5]
```diff
@@ -99,16 +101,24 @@ func (engineDetails EngineDetails) DeleteJobAccordingToJobCleanUpPolicy(experime
...
+		return jobCleanUpPolicy, fmt.Errorf("%v jobCleanUpPolicy not supported", jobCleanUpPolicy)
```
Comment (rahulchheda) [inline pkg/utils/watchJob.go:121]: `%s` maybe

File: pkg/utils/watchJob.go [hunk 6]
```diff
@@ -99,16 +101,24 @@ func (engineDetails EngineDetails) DeleteJobAccordingToJobCleanUpPolicy(experime
...
-		deleteJob := clients.KubeClient.BatchV1().Jobs(experiment.Namespace).Delete(experiment.JobName, &metav1.DeleteOptions{
+		if deleteJob := clients.KubeClient.BatchV1().Jobs(experiment.Namespace).Delete(experiment.JobName, &metav1.DeleteOptions{
 			PropagationPolicy: &deletePolicy,
-		})
-		if deleteJob != nil {
+		}); deleteJob != nil {
 			return "", errors.Errorf("unable to delete ChaosExperiment Job name: %v, in namespace: %v, error: %v", experiment.JobName, experiment.Namespace, err)
 		}
+	case v1alpha1.CleanUpPolicyRetain:
+		log.Infof("[skip]: skipping the job deletion as jobCleanUpPolicy is set to {%v}", expEngine.Spec.JobCleanUpPolicy)
```
Comment (rahulchheda) [inline pkg/utils/watchJob.go:119]: use `%s` directive for strings.

File: pkg/utils/watchJob.go [hunk 7]
```diff
@@ -99,16 +101,24 @@ func (engineDetails EngineDetails) DeleteJobAccordingToJobCleanUpPolicy(experime
...
 		deletePolicy := metav1.DeletePropagationForeground
-		deleteJob := clients.KubeClient.BatchV1().Jobs(experiment.Namespace).Delete(experiment.JobName, &metav1.DeleteOptions{
+		if deleteJob := clients.KubeClient.BatchV1().Jobs(experiment.Namespace).Delete(experiment.JobName, &metav1.DeleteOptions{
 			PropagationPolicy: &deletePolicy,
-		})
-		if deleteJob != nil {
+		}); deleteJob != nil {
 			return "", errors.Errorf("unable to delete ChaosExperiment Job name: %v, in namespace: %v, error: %v", experiment.JobName, experiment.Namespace, err)
 		}
+	case v1alpha1.CleanUpPolicyRetain:
```
Comment (rahulchheda) [inline pkg/utils/watchJob.go:118]: lets print some log when the `deleteJob == nil`

File: pkg/utils/environment.go (hunk 1 above)
Comment (ispeakc0de) [reply pkg/utils/environment.go:42]: - Added function to exclude all the envs who contains an empty value
-  as the number of envs increases the slice size increases which can fail bch guidelines(4x lines)

File: pkg/utils/watchJob.go (hunk 4 above)
Comment (ispeakc0de) [reply pkg/utils/watchJob.go:120]: it is added if someone added jobCleanUpPolicy to some wrong value other than delete or retain

File: pkg/utils/watchJob.go (hunk 7 above)
Comment (ispeakc0de) [reply pkg/utils/watchJob.go:118]: line 111 already printed for the positive case

**Review by ksatchit — APPROVED** [review]

**Review by uditgaurav — APPROVED** [review]

**Pull request merged**